### PR TITLE
Changing the grammar to support multiple top-level modules

### LINKED ISSUE #547: Support multiple modules
Currently, the parser parses just the first top-level module that appears in a file and ignores the rest. We should parse all of them, at least while we don't support multiple files.

### PR DESCRIPTION
**Pull request by konnov**
Closes #547. This PR partially addresses the issue of supporting multiple modules. I have extended the grammar and refactored the code to deal with multiple modules. However, there are a few places, where I am not sure about dealing with the type checker and the resolver.

@bugarela, could you have a look at the PR and maybe fix the problematic place in the type inference phase? Opening a draft PR, so we can fix it together, before merging the changes.

**Also, I broke the plugin integration.**

- [x] Entries added to the respective `CHANGELOG.md` for any new functionality

### CODE REVIEW

**Comment by bugarela** [comment]
On it!

**Comment by bugarela** [comment]
Ok, this is getting bigger and bigger. Our code is not prepared to compose results at this stage, so I'm having to change too many things. 

Here's what I'm going to do: on this PR, I'll fix the parsing phases so we support full parsing (with imports, conflict scanning and name resolution) for multiple modules; and fix the plugin to be compatible with that. 

On a follow up PR (or in multiple ones, if needed), I'll fix types, effects and modes.

It's ok to have partial support for multiple modules since right now we are ignoring those anyway, so it's not a regression.

**Comment by konnov** [comment]
Yeah, this is probably a bigger change than we expected. Let's just try to add multiple modules without removing nested modules first, as I would need multiple modules to replace nested module in REPL :)

**Comment by bugarela** [comment]
I think that's it for this first PR. @konnov shall we mark it as ready for review?

**Comment by bugarela** [comment]
Seems like CI is stuck somehow.

**Comment by konnov** [comment]
I have marked it ready for review. Thanks for fixing the code! We should keep an eye on multiple modules. They may hit us in unexpected places.

**Comment by konnov** [comment]
Hey @bugarela, do you think we should merge this one to unblock the other work such as #550? It would be good to have a review by @shonfeder. But since we know that this solution is interim, we can probably go ahead without Shon's review too.

**Review by shonfeder — APPROVED** [review]
i'm only part way thru, but will approve with my current notes so i don't blocke

File: quint/src/cliCommands.ts
Comment (bugarela) [reply quint/src/cliCommands.ts:164]: This is going to be fixed in a followup PR: https://github.com/informalsystems/quint/pull/559

File: quint/src/quintParserFrontend.ts
Comment (shonfeder) [inline quint/src/quintParserFrontend.ts:123]: if i am reading correctly, if `result` is a `left` here we just ignore the failure and continue working with the `module`? That's surprising to me. Also doesn't the `ParseResult<ParserPhase2>` already have the phase 1 modules? So what is the new `module` argument here for?

**Review by shonfeder — COMMENTED** [review]
still reviewing. sorry so slow. but have a few new questions.

File: quint/src/quintParserFrontend.ts
Comment (shonfeder) [inline quint/src/quintParserFrontend.ts:123]: I guess it's not introduced in this pr, the would be nice to do away with the mutable `definitions` that doesn't seem necessary afaict, but it does introduce a fair bit the complexity in trying to understand this function. Perhaps something for a follow up.

**Comment by shonfeder** [comment]
In general not sure how much help I can be and review here. The `parsePhase2` function has become pretty gnarly, in summary refactoring as a follow up might be helpful. But given the partially implemented nature of this, the context I'm missing on the general approach being applied, I'm not sure what I can offer in a time-efficient manner beyond the superficial remarks i've made.

File: quint/src/quintParserFrontend.ts
Comment (bugarela) [reply quint/src/quintParserFrontend.ts:104]: We haven't dropped support for nested modules yet

File: quint/src/quintParserFrontend.ts
Comment (bugarela) [reply quint/src/quintParserFrontend.ts:123]: I agree. That would make this diff significantly bigger, so it would fit better as a follow up. This is part of the first chunk of code me and Igor first wrote in typescript, and it deserves some care now that we have learned how to deal with situations. I have already refactor it to use `Either` (finally!), but it's still not as nice as our newest code.

File: quint/src/quintParserFrontend.ts
Comment (shonfeder) [reply quint/src/quintParserFrontend.ts:104]: ah, right. makes sense.

**Pull request merged**
